src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,448 @@
+    }
+    function _rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, y) / RAY;
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+
+        if (what == "vow") vow = data;
+        else if (what == "end") end = EndLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 ilk_, bytes32 what, uint256 data) external auth {
+        Ilk storage ilk = ilks[ilk_];
+        if (what == "tau" ) {
+            require(live == 1, "DssDirectDepositHub/hub-not-live");
+            require(ilk.tic == 0, "DssDirectDepositHub/join-not-live");
+
+            ilk.tau = data;
+        } else revert("DssDirectDepositHub/file-unrecognized-param");
+
+        emit File(ilk_, what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "pool") ilks[ilk].pool = D3MPoolLike(data);
+        else if (what == "vow") vow = data;
+        else if (what == "end") end = EndLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, D3MPoolLike pool, uint256 amount) internal {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        // Wind amount is limited by the debt ceiling
+        (uint256 Art,,, uint256 line,) = vat.ilks(ilk);
+        uint256 lineWad = line / RAY; // Round down to always be under the actual limit
+        if (_add(Art, amount) > lineWad) {
+            amount = _sub(lineWad, Art);
+        }
+
+        if (amount == 0) {
+            emit Wind(ilk, 0);
+            return;
+        }
+
+        require(int256(amount) >= 0, "DssDirectDepositHub/overflow");
+
+        uint256 sharesPrev = pool.shareBalance();
+
+        vat.slip(ilk, address(pool), int256(amount));
+        vat.frob(ilk, address(pool), address(pool), address(pool), int256(amount), int256(amount));
+        // normalized debt == erc20 DAI (Vat rate for this ilk fixed to 1 RAY)
+        pool.deposit(amount);
```

Comment:
What happened with the `daiJoin` usage?
It disappeared from here and in the `pool` is just used to get `dai()` address.
After `frob` it should be exiting it as erc20 before depositing in the pool.

Reply:
the `Pool` handles that since it is the owner of the balances. If I did it in the hub i would have a point where dai would pass through the hub just to be handed off to the pool which wasn't very clean

Reply:
https://github.com/makerdao/dss-direct-deposit/pull/31/commits/a6af7528bad745fccd7be1d460bc1f84591a3cb8